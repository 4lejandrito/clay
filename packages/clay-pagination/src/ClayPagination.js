import 'clay-link';
// eslint-disable-next-line
import { ClayDropdownBase } from 'clay-dropdown';
import {Config} from 'metal-state';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import templates from './ClayPagination.soy.js';

/**
 * Metal ClayPagination component.
 * @extends Component
 */
class ClayPagination extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayPagination.STATE = {
	/**
	 * Href to create each page complete href.
	 * @default undefined
	 * @instance
	 * @memberof ClayPagination
	 * @type {?(string|undefined)}
	 */
	baseHref: Config.string(),

	/**
	 * Current page number.
	 * @default undefined
	 * @instance
	 * @memberof ClayPagination
	 * @type {!number}
	 */
	currentPage: Config.number().required(),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayPagination
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayPagination
	 * @type {?(string|undefined)}
	 */
	id: Config.string(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayPagination
	 * @type {!string}
	 */
	spritemap: Config.string().required(),

	/**
	 * Total number of pages to show.
	 * @default undefined
	 * @instance
	 * @memberof ClayPagination
	 * @type {!number}
	 */
	totalPages: Config.number().required(),
};

defineWebComponent('clay-pagination', ClayPagination);

Soy.register(ClayPagination, templates);

export {ClayPagination};
export default ClayPagination;
