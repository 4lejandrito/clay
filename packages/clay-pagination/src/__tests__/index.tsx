/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import ClayPagination from '..';
import React from 'react';
import {
	cleanup,
	fireEvent,
	getByTestId,
	getByText,
	render,
} from '@testing-library/react';

const spritemap = 'path/to/spritemap';

describe('ClayPagination', () => {
	afterEach(cleanup);

	it('renders', () => {
		const {container} = render(
			<ClayPagination
				activePage={12}
				size="lg"
				spritemap={spritemap}
				totalPages={25}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('renders with only one page', () => {
		const {container} = render(
			<ClayPagination
				activePage={1}
				ellipsisBuffer={1}
				spritemap={spritemap}
				totalPages={1}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('calls onPageChange when arrow is clicked', () => {
		const changeMock = jest.fn();

		const {container} = render(
			<ClayPagination
				activePage={12}
				onPageChange={changeMock}
				spritemap={spritemap}
				totalPages={25}
			/>
		);

		fireEvent.click(
			getByTestId(container, 'prevArrow') as HTMLButtonElement,
			{}
		);

		expect(changeMock).toHaveBeenLastCalledWith(11);

		fireEvent.click(
			getByTestId(container, 'nextArrow') as HTMLButtonElement,
			{}
		);

		expect(changeMock).toHaveBeenLastCalledWith(13);
	});

	it('calls onPageChange when individual page is clicked', () => {
		const changeMock = jest.fn();

		const {container} = render(
			<ClayPagination
				activePage={12}
				onPageChange={changeMock}
				spritemap={spritemap}
				totalPages={25}
			/>
		);

		fireEvent.click(getByText(container, '25') as HTMLElement, {});

		expect(changeMock).toHaveBeenLastCalledWith(25);
	});

	it('shows dropdown when ellipsis is clicked', () => {
		const {getAllByText} = render(
			<ClayPagination
				activePage={12}
				spritemap={spritemap}
				totalPages={25}
			/>
		);

		fireEvent.click(getAllByText('...')[0] as HTMLElement, {});

		expect(
			document.body.querySelector('.dropdown-menu')!.classList
		).toContain('show');
	});

	it('calls onPageChange when iitem is clicked in dropdown-menu', () => {
		const changeMock = jest.fn();

		const {getAllByText} = render(
			<ClayPagination
				activePage={12}
				onPageChange={changeMock}
				spritemap={spritemap}
				totalPages={25}
			/>
		);

		fireEvent.click(getAllByText('...')[0] as HTMLElement, {});

		fireEvent.click(
			getByTestId(document.body, 'testId4') as HTMLElement,
			{}
		);

		expect(changeMock).toHaveBeenLastCalledWith(4);
	});
});
