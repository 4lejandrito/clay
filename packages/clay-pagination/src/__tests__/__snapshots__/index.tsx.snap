// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayPagination renders 1`] = `
<div>
  <ul
    class="pagination pagination-root"
  >
    <li
      class="page-item"
    >
      <button
        class="page-link"
        data-testid="prevArrow"
      >
        <svg
          class="lexicon-icon lexicon-icon-angle-left"
          role="presentation"
        >
          <use
            xlink:href="path/to/spritemap#angle-left"
          />
        </svg>
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        1
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        2
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        3
      </button>
    </li>
    <li
      class="dropdown page-item"
    >
      <span
        style="display: inline-block;"
      >
        <button
          class="dropdown-toggle page-link btn btn-unstyled"
          type="button"
        >
          ...
        </button>
      </span>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
      >
        25
      </button>
    </li>
    <li
      class="page-item"
    >
      <button
        class="page-link"
        data-testid="nextArrow"
      >
        <svg
          class="lexicon-icon lexicon-icon-angle-right"
          role="presentation"
        >
          <use
            xlink:href="path/to/spritemap#angle-right"
          />
        </svg>
      </button>
    </li>
  </ul>
</div>
`;

exports[`ClayPagination renders with only one page 1`] = `
<div>
  <ul
    class="pagination pagination-root"
  >
    <li
      class="page-item disabled"
    >
      <button
        class="page-link"
        data-testid="prevArrow"
        disabled=""
      >
        <svg
          class="lexicon-icon lexicon-icon-angle-left"
          role="presentation"
        >
          <use
            xlink:href="path/to/spritemap#angle-left"
          />
        </svg>
      </button>
    </li>
    <li
      class="page-item active"
    >
      <button
        class="page-link"
      >
        1
      </button>
    </li>
    <li
      class="page-item disabled"
    >
      <button
        class="page-link"
        data-testid="nextArrow"
        disabled=""
      >
        <svg
          class="lexicon-icon lexicon-icon-angle-right"
          role="presentation"
        >
          <use
            xlink:href="path/to/spritemap#angle-right"
          />
        </svg>
      </button>
    </li>
  </ul>
</div>
`;
